### PR TITLE
feat(repository): Add array decorator

### PR DESCRIPTION
**Pull request by kjdelisle**
### Description

Provide a nicer UX for the super-difficult-to-introspect array
properties, while limiting the amount of required information to
the class definition (while still allowing an optional props object)

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: packages/repository/src/decorators/model.ts [hunk 1]
```diff
@@ -73,3 +73,10 @@ export function property(definition?: Partial<PropertyDefinition>) {
     Object.assign({}, definition),
   );
 }
+
+export function array(
+  target: Function,
```
Comment (raymondfeng) [inline packages/repository/src/decorators/model.ts]: It should be `itemType`.

File: packages/repository/test/unit/decorator/model-and-relation.ts [hunk 2]
```diff
@@ -70,6 +85,8 @@ describe('model decorator', () => {
 
     // Validates that property no longer requires a parameter
     @property() isShipped: boolean;
+
+    @array(Product) items: Product[];
```
Comment (raymondfeng) [inline packages/repository/test/unit/decorator/model-and-relation.ts]: It might be better to name it as `@property.array` to be consistent with other decorators we have, such as:
- `@inject.getter`
- `@param.path.string`

Comment (raymondfeng) [inline packages/repository/test/unit/decorator/model-and-relation.ts]: We should validate to make sure the decorated property has design-time type to be `Array` even the item type is not captured by TypeScript at the moment.

File: packages/repository/test/unit/decorator/model-and-relation.ts [hunk 3]
```diff
@@ -116,6 +127,7 @@ describe('model decorator', () => {
     });
     expect(meta.id).to.eql({type: 'string', id: true, generated: true});
     expect(meta.isShipped).to.eql({type: Boolean});
+    expect(meta.items).to.eql({type: Product, array: true});
```
Comment (b-admike) [inline packages/repository/test/unit/decorator/model-and-relation.ts]: Can you please explain why here `type` is `Product` which is a Class, but https://github.com/strongloop/loopback-next/pull/806/files#diff-aa9599539cae81d1bbc3cdb480ecdb09R81 we define it as a function?

Comment (kjdelisle) [reply packages/repository/test/unit/decorator/model-and-relation.ts]: All objects are Functions, and this is the only way we can properly handle class constructors in a way that doesn't resort to the use of `any`. If you look at most of our other decorators, we use the same approach.

Comment (b-admike) [reply packages/repository/test/unit/decorator/model-and-relation.ts]: Thanks 👍

File: packages/repository/src/decorators/model.ts [hunk 4]
```diff
@@ -8,9 +8,11 @@ import {
   ClassDecoratorFactory,
   PropertyDecoratorFactory,
   MetadataMap,
+  Constructor,
 } from '@loopback/context';
 import {ModelDefinition, ModelDefinitionSyntax} from '../model';
 import {PropertyDefinition} from '../index';
+import {ItemType} from '../../../openapi-spec/index';
```
Comment (raymondfeng) [inline packages/repository/src/decorators/model.ts]: This import is problematic.

File: packages/repository/src/decorators/model.ts [hunk 5]
```diff
@@ -73,3 +75,26 @@ export function property(definition?: Partial<PropertyDefinition>) {
...
+  export const array = function(
+    itemType: Function,
+    definition?: Partial<PropertyDefinition>,
+  ) {
+    return function(target: Object, propertyName: string) {
+      const propType = MetadataInspector.getDesignTypeForProperty(
+        target,
+        propertyName,
+      );
+      if (!Array.isArray(propType())) {
```
Comment (raymondfeng) [inline packages/repository/src/decorators/model.ts]: The test should be `propType !== Array`.

Comment (kjdelisle) [reply packages/repository/src/decorators/model.ts]: It wasn't letting me make that comparison before, though, that may have been due to another incorrect type. Let me try again.

**Comment by raymondfeng** [comment]
LGTM. Since you are here, please also fix the following imports in `model.ts`:
```ts
import {ModelDefinition, ModelDefinitionSyntax} from '../model';
import {PropertyDefinition} from '../index';
```

The consistent one should be:
```ts
import {ModelDefinition, ModelDefinitionSyntax, PropertyDefinition} from '../model';
```

File: packages/repository/src/decorators/model.ts [hunk 6]
```diff
@@ -73,3 +73,28 @@ export function property(definition?: Partial<PropertyDefinition>) {
     Object.assign({}, definition),
   );
 }
+
+export namespace property {
+  export const ERR_PROP_NOT_ARRAY =
+    '@property.array can only decorate array properties!';
+  export const array = function(
```
Comment (raymondfeng) [inline packages/repository/src/decorators/model.ts]: It would be better to add tsdoc for this decorator function.
